US81119 - Restyle and change search bar's internal behaviour

ryantmer opened the pull request:
The search bar as it was previously did not match the behaviour of other search bars in the LMS. In particular, the desired behaviour is such that the state switches between "searched" and "not searched". This is effectively reflected through the state of the button icon - if you have searched for your current term, the button (X) will clear the field, and "redo the search with no query", clearing it back to the "not searched" state. If you change the term (or, in the initial state), the button goes back to the magnifying glass, and activating it (or hitting Enter) does the query and enters the "searched" state.

This also involved some rework of the under-the-hood stuff here, to take fuller advantage of some of Polymer's features as well as address some minor issues that previous existed.

Fixes #247 (kind of a side effect)
Fixes #248 (kind of a side effect)

"While I was here" I also added caching of enrollments search responses, which fixes #245.

This will also close #246, since the search as is done in the LMS doesn't follow that same paradigm - when you use backspace to clear a search field in the LMS, it doesn't re-search with the empty query; either you have to hit Enter, or you can clear it by clicking the X button.

See also https://github.com/Brightspace/d2l-search-widget-ui/pull/13 - a related change (that component already had live search, so it's really just a UI/slight UX change there).

ctwomey1 commented on d2l-search-widget-custom.html:
```diff
@@ -315,84 +288,104 @@
+				} else {
+					this._fullSearch();
+					this._setClearIcon();
 				}
```
Depending on the icons file name to determine actions seems a bit odd to me, would it make more sense to toggle what the on tap does in the _clearSearch and _fullSearch functions

ryantmer replied:
Yeah, I'm not totally sold on this approach, but the alternative was keeping an internal state of "is currently searched"/"is currently cleared". I figured that information was redundant to the state of the icon, but I think you're right, it is more clear that way.

ctwomey1 commented on d2l-search-widget-custom.html:
```diff
@@ -315,84 +288,104 @@
 
-			_onSearchStringChanged: function(newString) {
+			_setClearIcon: function() {
+				this.$$('d2l-icon').setAttribute('icon', 'd2l-tier1:close-default');
```
This is a pretty loose querySelector, you might want to be more specific (ie like adding a id/class or doing something like this.$$('div.search-bar d2l-icon') because if someone added  d2l-icon higher up in the template i believe that this would grab that instead of your intended element.

if you tighten up the querySelector you should do a pass through this file and correct all of them

ryantmer commented:
Argh, yeah forgot tests. I got to the end of the day, committed, and moved to something else the next day. I'll switch this back to WIP while I'm working on them. Thanks!

ryantmer commented:
Hokay, added a couple tests, tightened the querySelector, changed my keyboard backlight colour, and solidified the "searched"/"not searched" state representation. I _think_ the change I made in that regard makes things a bit less convoluted, hence perhaps fewer tests than expected. I tested the three cases for icon-changing that I was playing with the UI while writing this, notably:

- searching for something (button click/Enter) should do the search and change the icon to the "X" icon
- clearing after searching should re-fetch the (cached irl, not-cached in the test) response and change to the "search" icon
- searching for something, then changing the search term should immediately go back to the "search" icon (not redo the search until button click/Enter though)

The first two use the `_updateIcon` call after `_isSearched` has been changed; unfortunately the third case requires calling the change to the icon directly (since `_isSearched === true` in that case, but we also want the search icon, not the clear icon).

JoshuaKlassen commented:
These are my notes from testing:

Testing notes:
	
[X] Icon/button is removed from left side of field
	-All browsers passed
[X] With an empty search field, the Search icon is right-aligned in the search field
	-All browsers passed
[] When a user types, no changes occurs
	-Internet Explorer and Edge failed
		-On the All Courses search widget, the X icon appears to the left of the Search Icon when typing
[] When a user clicks the search icon (or presses Enter), the search happens and the icon changes to the X icon
	-Failed on all browsers
		-The filter search widgets don't change the Icon when clicking, but work with pressing enter
[] When the X icon is clicked, the search field is cleared, and results are updated back to the results for an empty search/no search
	-Failed on all browsers
		-With the above criteria failing, I could only test with pressing enter first, then clicking the icons to get the X icon
			-Clicking the X icon does clear the search, but does not change the icon back to the Search Icon
	
Mobile:
Chrome dev tools mobile:
	-Same issues as desktop browser

iOS (Safari):
	-Same issues as desktop browser

Accessibility (NVDA Screen Reader):
	-Tabbing works
	
	-Not sure if the screen reader would original read out results here, but it doesn't now
	
Internet Explorer Notes:
	-The default behaviour in Internet Explorer is that clicking in the box udpates the Icon to the X Icon
	
	-The two search widgets in Filter (Semesters and Departments) appear stacked at first with the X Icon, and are not able to perform searchs, until either the Semester tab or the Department tab is clicked
		-Also when searching for 'Math' in the Departments filter, clearing it does something weird with the original results - the number of rows shrinks in that panel so I can't see all the departments
		
Edge notes:
	-Has the same issues as Internet Explorer with the two filter searches stacking
	
Other notes:
	-I noticed that the Magnifying Glass is flipped with respect to the `/d2l/le/content/123456/Home` Icon
		-I don't know how closly these search widgets are supposed to match the other widgets on the LMS
	
	-I noticed a defect(?) on mobile: Clicking the 'Filter' label opens and immediately closes the dropdown, I had to click the arrow to open it.
	
	-We can do a Zoom meeting to discuss some of the issues if they are confusing to understand - I can share my screen and show you

ryantmer commented:
Okay, I believe I addressed the three outstanding items you mentioned there - I had made some changes here, but forgot to make the same changes in `d2l-search-widget` (but that PR has now been updated). Also fixed the extra X that IE adds.

For the IE/Edge notes, I'll have to take a look tomorrow; not sure what in here would cause that change :S (does it do it on master, or only with this PR checked out?)

For the other notes, the flipped icon I presume is okay, although I hadn't noticed that... if anything, I'd argue the one in the LMS is backwards 😛 As for the filter menu on mobile, I believe we decided not to make it mobile-friendly (time constraints yay), although it's odd that it's showing up at all... Once I've got all my stuff back up and running I'll reach out to you if I need any more clarifications about the outstanding stuff. Thanks!

JoshuaKlassen commented:
I only tested with this PR checked out, I will see what master does in IE/Edge!

ryantmer commented:
Ah, okay - this PR alone will do the things for the All Courses search, but you'll need the search-widget PR for the Filter menu search boxes.

JoshuaKlassen commented:
Testing notes:

---------------------------------------------
Acceptence Criteria:
- [X] Icon/button is removed from left side of field

- [X] With an empty search field, Search icon is right-aligned in the search field 

- [X] When a user types, no changes occur 

- [X] When a user clicks the search icon (or presses Enter), the search happens and the icon changes to the X Icon 

- [X] When the X icon is clicked, the search field is cleared, and results are updated back to the results for an empty search/no search 

---------------------------------------------
Tested on:
- [X] Chrome 

- [X] Firefox 

- [X] Internet Explorer 

- [X] Microsoft Edge

- [X] Safari

- [X] Android (Chrome dev tools mobile view)

- [X] iOS (iPad)

---------------------------------------------
Other notes:
-The issue with the extra X icon appearing when typing in All Courses in IE/Edge no longer happens! (fixed)

-Edge no longer stacks the filter searches first time opening the dropdown (IE still does) 

-The issue with the list of departments shrinking happens in Firefox, IE, Edge, and Safari (I missed Firefox and Safari last time)

-There is an issue in the All Courses search bar, where after a search, if I delete a character (not the whole search string), the icon changes back to the 'Search Icon' (expected), but clicking it acts as the 'X' icon, cleaing the search bar and updating the results back to the results of an empty search.

-When typing in a filter search bar(Semester or Department), clicking inside the All Courses search bar gives mouse focus, but not keyboard focus, have to click inside again to fix. This exists in all browsers.

ryantmer commented:
Sweet, thanks! w.r.t. the other notes:

- The department/semester lists shrinking is happening on master, so wasn't introduced here looks like. That said, I suspect it's related to #168 - I've added a note there to look out for fixing that if we ever figure out wtf is causing that
- The weird focus thing also wasn't introduced here, but good catch! I've added a GH issue for that - it's pretty minor, but is the sort of thing that shouldn't be forgotten. See #279.
- [ ] Search button acting as a clear is definitely something I introduced here, I'll fix that up right meow

ryantmer commented:
Hokay, fixed the one outstanding issue, should be good now :)

JoshuaKlassen commented:
It looks like filter search bars now don't switch to the search icon, it stays the X icon (in Chrome)

ryantmer commented:
Oh dammit... that would be because I made the change here, but not in the dependency repo. Note to self: after this PR, move to a common behavior as @ctwomey1 very intelligently suggested...

ryantmer commented:
Updated the other repo and bumped the version here, should be good now. Sorry for all the back and forth! D:

JoshuaKlassen commented:
Everything looks good!
